pep-0458.txt

```diff
@@ -730,99 +735,97 @@ efficiently transfer consistent snapshots from PyPI.
-every metadata and target file (except for *timestamp* metadata) by including,
-in the file request, the version of the file (for metadata), or the
-cryptographic hash of the file (for target files) in the filename.
-
-Finally, PyPI SHOULD use a `transaction log`__ to record project transaction
-processes and queues so that it will be easier to recover from errors after a
-server failure.
+When a new project release is uploaded to PyPI, PyPI MUST update the *bin-n*
+metadata responsible for the target files of the project release (remember that
+target files are sorted into bins by their filename hashes). Consequentially,
+PyPI MUST update *snaphot* to account for the updated *bin-n* metadata, and
+*timestamp* to account for the updated *snapshot* metadata. These updates
+SHOULD be handled by automated processes, e.g. one or more *transaction
+processes* and one *snapshot process*.
+
+The transaction process keeps track of a project upload, adds all new target
+files to the most recent, relevant *bin-n* metadata (in memory) and informs the
+snapshot process to produce a consistent snapshot. Each project release SHOULD
+be handled in an atomic transaction, so that a given consistent snapshot
+contains all target files of a project release. However, transaction processes
+MAY be parallelized under the following constraints:
+
+- No pair of transaction processes MUST concurrently work on the same project.
+- No pair of transaction processes MUST concurrently work on projects that
+  belong to the same *bin-n* role.
+
+When a transaction process is finished updating the relevant *bin-n* metadata
+it informs the snapshot process to generate a new consistent snapshot. The
+snapshot process does so by taking the updated *bin-n* metadata, incrementing
+their respective version numbers, signing them with the *bin-n* role key(s),
+and writing them to *VERSION_NUMBER.bin-N.json*.
+
+Similarly, the snapshot process then takes the most recent *snapshot* metadata,
+updates its *bin-n* metadata hashes and version numbers, increments its own
+version number, signs it with the *snapshot* role key, and writes it to
+*VERSION_NUMBER.snapshot.json*.
+
+And finally, the snapshot process takes the most recent *timestamp* metadata,
+updates its *snapshot* metadata hash and version number, increments its own
+version number, sets a new expiration time, signs it with the *timestamp* role
+key, and writes it to *timestamp.json*.
+
+The snapshot process MUST generate consistent snapshots sequentially, reading
+the notifications received from the transaction process(es) from a
+concurrency-safe FIFO queue. Fortunately, the operation of signing is fast
+enough that this may be done a thousand or more times per second.
+
+The following caveats exist in regards to project transactions and immediate
+availability of target files:
+
+- PyPI allows adding files to a project release at arbitrary later times.
```

Comment:
What does this mean?

Reply:
Ernest mentioned this in https://github.com/python/peps/pull/1203#discussion_r336229332:
> PyPI allows for upload of additional files for a release at any point in the future

I can use his wording, if you prefer.

Reply:
The original is clearer, though I can improve it a bit.
```suggestion
- PyPI allows permits additional  files to be uploaded to a project release at any point in the future.
```